syntax = "proto3";
package t4proto.v1.auth;

import "google/protobuf/timestamp.proto";
import "t4/v1/common/enums.proto";

message LoginRequest {
  // API Key authentication (when used, this is the only required field)
  string api_key = 1;
  
  // Username/password authentication fields
  string firm = 2;
  string username = 3;
  string password = 4;
  string app_name = 5;
  string app_license = 6;
  t4proto.v1.common.PriceFormat price_format = 10;
}

message LoginResponse {
  t4proto.v1.common.LoginResult result = 1;
  string session_id = 2;
  string user_id = 3;
  string firm_id = 4;
  repeated string roles = 5;
  string error_message = 6;

  message Exchange {
    string exchange_id = 1;
    string user_id = 2;
    t4proto.v1.common.MarketDataType market_data_type = 3;
    bool has_executing_account = 4;
    string primary_user_id = 5;
    string secondary_user_id = 6;
    string location = 7;
    string smp_id = 8;
    map<string, string> extra_detail = 9;
  }

  message Account {
    string account_id = 1;
    string account_number = 2;
    string account_name = 3;
    string display_name = 4;
    t4proto.v1.common.AccountMode mode = 5;
  }

  repeated Exchange exchanges = 7;
  repeated Account accounts = 8;

  optional AuthenticationToken authentication_token = 9;
}

message AuthenticationTokenRequest {
  string request_id = 1;
}

message AuthenticationToken {
  string request_id = 1;
  optional string token = 2;
  optional google.protobuf.Timestamp expire_time = 3;
  optional string fail_message = 4;
}